### PR TITLE
feat: add Emails on Push service (#607)

### PR DESCRIPTION
**Pull request by christian-sahlmann**

### CODE REVIEW

**Comment by timofurrer** [comment]
@christian-sahlmann thanks for the contribution 🎉  Could you please rebase onto the latest main (see #1304)?

File: go.mod [hunk 1]
```diff
@@ -8,7 +8,7 @@ require (
 	github.com/hashicorp/terraform-plugin-sdk/v2 v2.19.0
 	github.com/mitchellh/hashstructure v1.1.0
 	github.com/onsi/gomega v1.22.1
-	github.com/xanzy/go-gitlab v0.73.1
+	github.com/xanzy/go-gitlab v0.74.0
```
Comment (timofurrer) [inline go.mod:11]: `go-gitlab@v0.74.0` will be integrated with https://github.com/gitlabhq/terraform-provider-gitlab/pull/1308. Thus, another rebase may be required ;)

File: internal/provider/resource_gitlab_service_emails_on_push.go [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+			"tag_push_events": {
+				Description: "Enable notifications for tag push events.",
+				Type:        schema.TypeBool,
+				Optional:    true,
+			},
+			"branches_to_be_notified": {
+				Description: `Branches to send notifications for. Valid options are "all", "default", "protected", and "default_and_protected". Notifications are always fired for tag pushes.`,
+				Type:        schema.TypeString,
+				Optional:    true,
+				Default:     "all",
```
Comment (timofurrer) [inline internal/provider/resource_gitlab_service_emails_on_push.go:66]: 🚧 **Suggestion:** let's add a validator here, similar to what we have here:

https://github.com/gitlabhq/terraform-provider-gitlab/blob/96e5f299d937aa1d74281aa21cbb92938f286cbe/internal/provider/resource_gitlab_group_access_token.go#L76

Also make sure to render the options accordingly in the `Descriptions`. Also see the above example.

File: internal/provider/resource_gitlab_service_emails_on_push.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+				Computed:    true,
+			},
+		},
+	}
+})
+
+func resourceGitlabServiceEmailsOnPushCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*gitlab.Client)
+	project := d.Get("project").(string)
+	d.SetId(project)
```
Comment (timofurrer) [inline internal/provider/resource_gitlab_service_emails_on_push.go:100]: 🚧 **Suggestion:** let's only set the `id` after we actually have a successfully created integration.

File: internal/provider/resource_gitlab_service_emails_on_push.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+	options := &gitlab.SetEmailsOnPushServiceOptions{
+		Recipients:             gitlab.String(d.Get("recipients").(string)),
+		DisableDiffs:           gitlab.Bool(d.Get("disable_diffs").(bool)),
+		SendFromCommitterEmail: gitlab.Bool(d.Get("send_from_committer_email").(bool)),
+		PushEvents:             gitlab.Bool(d.Get("push_events").(bool)),
+		TagPushEvents:          gitlab.Bool(d.Get("tag_push_events").(bool)),
+		BranchesToBeNotified:   gitlab.String(d.Get("branches_to_be_notified").(string)),
```
Comment (timofurrer) [inline internal/provider/resource_gitlab_service_emails_on_push.go:108]: 🚧 **Suggestion:** for better compatibility we usually leave the defaults up to the GitLab API. Just setting all the fields in the request may lead to incpompatibilies between the defaults in the provider and the API.

You can use the following pattern to construct the options:

https://github.com/gitlabhq/terraform-provider-gitlab/blob/96e5f299d937aa1d74281aa21cbb92938f286cbe/internal/provider/resource_gitlab_cluster_agent_token.go#L45

File: internal/provider/resource_gitlab_service_emails_on_push.go [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+		ReadContext:   resourceGitlabServiceEmailsOnPushRead,
+		UpdateContext: resourceGitlabServiceEmailsOnPushCreate,
+		DeleteContext: resourceGitlabServiceEmailsOnPushDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"project": {
+				Description:  "ID of the project you want to activate integration on.",
```
Comment (timofurrer) [inline internal/provider/resource_gitlab_service_emails_on_push.go:30]:
```suggestion
				Description:  "ID or full-path of the project you want to activate integration on.",
```

File: internal/provider/resource_gitlab_service_emails_on_push.go [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+func resourceGitlabServiceEmailsOnPushRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*gitlab.Client)
+	project := d.Id()
+
+	log.Printf("[DEBUG] read gitlab emails on push service for project %s", project)
+
+	service, _, err := client.Services.GetEmailsOnPushService(project, gitlab.WithContext(ctx))
+	if err != nil {
+		if is404(err) {
+			log.Printf("[DEBUG] gitlab emails on push service not found for project %s", project)
```
Comment (timofurrer) [inline internal/provider/resource_gitlab_service_emails_on_push.go:130]: 💡 **(non-blocking) Nitpick:** 

```suggestion
			log.Printf("[DEBUG] gitlab emails on push service not found for project %s, removing from state", project)
```

File: internal/provider/resource_gitlab_service_emails_on_push_test.go [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+			{
+				ResourceName:      "gitlab_service_emails_on_push.this",
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+		},
+	})
+}
+
+func testAccCheckGitlabServiceEmailsOnPushExists(resourceIdentifier string, service *gitlab.EmailsOnPushService) resource.TestCheckFunc {
```
Comment (timofurrer) [inline internal/provider/resource_gitlab_service_emails_on_push_test.go:97]: I don't think this adds any value. The import test step will actually ensure it works properly. 

See here https://github.com/gitlabhq/terraform-provider-gitlab/issues/1008#:~:text=and%20data%20sources.-,Use%20ImportStateVerify%20TestSteps%20as%20a%20replacement%20for%20writing%20custom%20%22CheckExists%22%20and%20%22CheckAttributes%22%20check%20funcs.,-Undesirable%3A

WDYT?

File: internal/provider/resource_gitlab_service_emails_on_push_test.go [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+	var recipients2 = "mynumbertworecipient@example.com"
+	var emailsOnPushResourceName = "gitlab_service_emails_on_push.this"
+
+	resource.ParallelTest(t, resource.TestCase{
+		ProviderFactories: providerFactories,
+		CheckDestroy:      testAccCheckGitlabServiceEmailsOnPushDestroy,
+		Steps: []resource.TestStep{
+			// Create an Emails on Push service
+			{
+				Config: testAccGitlabServiceEmailsOnPushConfig(testProject.ID, recipients1),
```
Comment (timofurrer) [inline internal/provider/resource_gitlab_service_emails_on_push_test.go:31]: We started inlining the configurations for better readability. Thus, could you please inline it?

See https://github.com/gitlabhq/terraform-provider-gitlab/issues/1008#:~:text=Prefer%20writing%20config%20in%2Dline%20to%20wrapping%20it%20in%20a%20function.

**Review by timofurrer — CHANGES_REQUESTED** [review]
@christian-sahlmann already looks pretty solid. I've left a few comments 🏓

File: go.mod (hunk 1 above)
Comment (christian-sahlmann) [reply go.mod:11]: Okay, I'll keep an eye on it and do another rebase in case #1308 gets merged before this :)

Comment (timofurrer) [reply go.mod:11]: It's merged btw :)

File: internal/provider/resource_gitlab_service_emails_on_push.go (hunk 2 above)
Comment (christian-sahlmann) [reply internal/provider/resource_gitlab_service_emails_on_push.go]: Cool, I was suspecting there is a nicer way to describe this :) Thanks for pointing out

File: internal/provider/resource_gitlab_service_emails_on_push.go (hunk 3 above)
Comment (christian-sahlmann) [reply internal/provider/resource_gitlab_service_emails_on_push.go]: makes sense, done

File: internal/provider/resource_gitlab_service_emails_on_push.go (hunk 4 above)
Comment (christian-sahlmann) [reply internal/provider/resource_gitlab_service_emails_on_push.go]: cool, didn't know about this. Adapted the code accordingly

File: internal/provider/resource_gitlab_service_emails_on_push_test.go (hunk 7 above)
Comment (christian-sahlmann) [reply internal/provider/resource_gitlab_service_emails_on_push_test.go:108]: yes, looking at it, it seems to be already fully covered by ImportVerify

**Comment by christian-sahlmann** [comment]
@timofurrer, thanks for the great review and for being gentle with a go beginner ;) I hope I adressed all comments properly.

**Review by timofurrer — APPROVED** [review]
@christian-sahlmann thanks for the fast update - it looks ready to merge 🎉 

... and you did great - especially as a Go beginner 💪 keep it up 🌮

**Pull request merged**